Doc/regions.md

```diff
@@ -0,0 +1,517 @@
+Here we present the idea as a reverse polish evaluator.
+```python
+>>> x = 1
+>>> y = 2
+>>> calc("4 x y + *")
+```
+The idea is to interpolate the `x` and `y` into the string while evaluating it.
+```
+def calc(str):
+  ctxt = sys._getframe().f_back
+  ops = str.split()
+  stack = []
+  for op in ops:
+    try: 
+      stack.append(int(op))
+    except ValueError:
+      if op == "+":
+        stack.append(stack.pop() + stack.pop())
+      elif op == "*":
+        stack.append(stack.pop() * stack.pop())
+      else:    
+        stack.append(ctxt.f_locals[op])      
+  return stack.pop()
+```
+This is a simple example, but it shows how the frame object can be used to access the local variables in the calling scope.
+```
+>>> x = 1
+>>> y = 2
+>>> calc("4 x y + *")
+12
+```
+
+The numpy library is pretty popular, so this is a good example of how to use the frame object in a way that is not debugging.
+
+
+Note that the frame object is mostly read-only, but does contain an interesting operation to set the executing line no of a frame.
+```C
+static int
+frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno, void *Py_UNUSED(ignored))
+```
+This seems very bad to call from another thread.
+
+Ultimately, a frame object should never escape the thread that it was created in.
+
+
+
+## Regions in Veronapy.
+
+The previous section illustrates the many temporary objects that Python creates to handle variables.
+To support this, we need our region object model to support complex local structures for indexing into regions.
+To achieve this, we will use three kinds of object ownership:
+
+* Local object 
+* Immutable
+* Regions
+
+The "local objects" are used to represent the many local structures that Python creates such as function objects, frame objects and cells.
+These objects are typically not shared between threads, so normally should not require any region tracking.
+However, as it is possible for them to become shared, we need to be able to make them "immutable".
+
+To pass mutable states between threads, we need to introduce a new construct to Python: "regions".
+A region can be accessed by at most one thread at a time.
+This removes the need for locks to protect the data structure.
+The region can be passed between threads, and the receiving thread can access the data structure.
+
+To efficiently support regions, we impose the following invariants that the runtime is responsible for maintaining:
+
+* Local objects can reach anything
+* Immutable objects can only reach other immutable objects.
+* Objects in a region can only reach other objects in the same region, immutable objects, or linear region entry points.
+* Regions track the number of incoming references from local objects.
+
+New objects are always created as a local object.
+If Veronapy features are not used, then all the objects in the system will be local objects and the runtime will behave as normal.
```

Comment:
Minor: While I agree with this, because we talk about threads above, and non-Veronapy Python programs can use Python threads, someone might read this wrong. All objects are local objects so they should not be shareable across threads… yet the runtime behaves as normal?

Reply:
I wonder if I should use "interpreter" instead of "thread".  Then local objects are local to an interpreter. To move them between interpreters, they must either be in regions or immutable.